analytics/http/filters.go

```diff
@@ -0,0 +1,189 @@
+package http
+
+import (
+	"math/rand"
+
+	"github.com/antonmedv/expr"
+	"github.com/antonmedv/expr/vm"
+	"github.com/prebid/prebid-server/v2/analytics"
+	"github.com/prebid/prebid-server/v2/config"
+)
+
+type (
+	auctionFilter      func(event *analytics.AuctionObject) bool
+	ampFilter          func(event *analytics.AmpObject) bool
+	cookieSyncFilter   func(event *analytics.CookieSyncObject) bool
+	notificationFilter func(event *analytics.NotificationEvent) bool
+	setUIDFilter       func(event *analytics.SetUIDObject) bool
+	videoFilter        func(event *analytics.VideoObject) bool
+)
+
+func createAuctionFilter(feature config.AnalyticsFeature) (auctionFilter, error) {
+	var filterProgram *vm.Program
+	var err error
+	if feature.Filter != "" {
+		// precompile the filter expression for performance, make sure we return a boolean from the expression
+		filterProgram, err = expr.Compile(feature.Filter, expr.Env(analytics.AuctionObject{}), expr.AsBool())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return func(event *analytics.AuctionObject) bool {
+		// Disable tracking for nil events or events with a sample rate of 0
+		if event == nil || feature.SampleRate <= 0 || rand.Float64() > feature.SampleRate {
```

Comment:
Please abstract away `rand` with the `randomutil`, just as `clock` is an abstraction, for testing purposes.